grakn-kb/src/main/java/ai/grakn/kb/internal/concept/RelationshipReified.java

```diff
@@ -58,6 +65,19 @@
  *
  */
 public class RelationshipReified extends ThingImpl<Relationship, RelationshipType> implements RelationshipStructure {
+    /**
+     * Set of {@link Casting}s which are loaded into memory. This should only be used when adding new roleplayers.
+     * This is because validation requires iterating over all castings anyway.
+     */
+    private Cache<Set<Casting>> allCastings = new Cache<>(Cacheable.set(), () -> {
+        return vertex().tx().getTinkerTraversal().V().
+                has(Schema.VertexProperty.ID.name(), getId().getValue()).
+                outE(Schema.EdgeLabel.ROLE_PLAYER.getLabel()).
+                has(Schema.EdgeProperty.RELATIONSHIP_TYPE_LABEL_ID.name(), type().getLabelId().getValue()).
+                toStream().map(edge -> vertex().tx().factory().buildCasting(edge)).
+                collect(Collectors.toSet());
```

Comment:
Any reason we can't just do `vertex().getEdgesOfType(Direction.OUT, Schema.EdgeLabel.ROLE_PLAYER)`?

Reply:
If I did that then the index is ignored. I will put a comment there.